### PR TITLE
Avoiding dependency on EnsoContext from Module and ModuleScope

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Less dependencies on `EnsoContext` - `Module` shall exist without it. `Module` is a result of a `Compiler` and shall be created before its execution - e.g. requiring `EnsoContext` (with all its runtime information) is a bit too _demanding_.

### Important Notes

The only reason why `Module` wanted `EnsoContext` was to create its (associated) `Type`. `Type`'s constructor needed a parent type and the code was asking for `Any` from the context. That's unnecessary at creation time - we can just use some constant (like `null`) and turning it into `Any` during execution. Benchmarks show that there is no slowdown doing so.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- All code has been tested:
  - [x] Unit tests continue to pass
  - [x] Benchmark results were checked

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
There is a [failure in RuntimeVizualizationsTest](https://github.com/enso-org/enso/actions/runs/5540956481/jobs/10113720356?pr=7277#step:10:6323). Investigations shows that there is a difference in `getSupertype()` of `Error` - originally it returned `null`, but now it returns `Any` - I assume it should return `Any` - so maybe the test needs a fix. CCing @4e6.

Is it correct that [class Error extends Builtin](https://github.com/enso-org/enso/blame/b042807790da16b0bde626ae8c9c3ed9ac63179c/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/Error.java#L9) overrides method `getSuperType()` and returns `null`, @kustosz? I believe it is wrong - `Error` is also a subclass of `Any`, right?

**Marcin's reply**: No, it is intentional _Error isn't subclass of Any!_ - adjusted in 934bbed - new test written. Code changed for builtins to return `null` super type.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java [hunk 1]
```diff
@@ -118,6 +118,13 @@ public boolean isBuiltin() {
   }
 
   public Type getSupertype() {
+    if (supertype == null) {
+      if (builtin) {
+        return null;
+      }
+      var ctx = EnsoContext.get(null);
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java:125]: Calling `get(null)` might affect a performance a bit. Running [benchmarks to verify](https://github.com/enso-org/enso/actions/runs/5546262974).

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java [hunk 2]
```diff
@@ -32,22 +31,16 @@ public final class ModuleScope implements TruffleObject {
...
     this.imports = new HashSet<>();
     this.exports = new HashSet<>();
     this.module = module;
-    this.associatedType =
-        Type.createSingleton(
-            module.getName().item(),
-            this,
-            context == null ? null : context.getBuiltins().any(),
-            false);
+    this.associatedType = Type.createSingleton(module.getName().item(), this, null, false);
```
Comment (4e6) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:43]: Not sure about that. Isn't it breaking the type hierarchy? That Any is a supertype of all types

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:43]: The type hierarchy remains unchanged. This PR:
- trades construction time for time spent in `getSupertype()`
- ensures only builtins (`Any` and `Error` currently) can have no parent
- all other types either specify their parent or get `Any` in `getSupertype()`

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java:125]: Benchmarks seem to be unaffected by this change. Thanks @Akirathan for the new benchmarking tool.

Sample benchmark comparition

**Review by 4e6 — APPROVED** [review]

**Comment by JaroslavTulach** [comment]
[TextExtensionsTest failure](https://github.com/enso-org/enso/actions/runs/5545898079/jobs/10157809554?pr=7277#step:10:6597).

**Pull request merged**